### PR TITLE
[ALLUXIO-2815] Use AlluxioStatusException throughout client

### PR DESCRIPTION
**Pull request by aaudiber**
https://alluxio.atlassian.net/browse/ALLUXIO-2815

1. Add the capability for `AlluxioStatusExceptions` to wrap `IOExceptions`. We have the property that `AlluxioStatusException.fromIOException(e).toIOException() == e` as long as `e instanceof IOException`.
2. Within the boundary of our client public API, all exceptions we throw are `AlluxioStatusExceptions`.
3. At the public API borders (`BaseFileSystem`, `AbstractFileSystem`, `FileInStream`, `FileOutStream`, `AbstractLineageClient`, `LineageFileSystem`), we convert any `AlluxioStatusExceptions` to the exceptions used by our public API. No `AlluxioStatusExceptions` should escape the boundary of our public API. In 2.0 we will remove the current public API exceptions and either throw a checked version of `AlluxioStatusException` or directly throw `AlluxioStatusException`.
4. Many places where we would catch checked exceptions previously were actually incorrect in the face of unchecked exceptions. This has been fixed.
5. Update `IOException` javadoc to say `@throws IOException if an I/O error occurs`. Given the nature of `IOException`, any more detailed message we provide can very easily become incorrect.

### CODE REVIEW

**Comment by aaudiber** [comment]
@calvinjia @apc999 @jsimsa This PR is ready for review

**Comment by aaudiber** [comment]
Let me know if you'd like the files to be split into multiple sub-PRs

File: core/client/src/main/java/alluxio/client/AbstractOutStream.java [hunk 1]
```diff
@@ -40,7 +40,7 @@ public int getBytesWritten() {
   /**
    * Aborts the output stream.
    *
-   * @throws IOException if there is a failure to abort this output stream
+   * @throws IOException if an I/O error occurs
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/AbstractOutStream.java]: This seems redundant, is there suggested javadoc style for cases like this?

File: core/client/src/main/java/alluxio/client/FileSystemTestUtils.java [hunk 2]
```diff
@@ -37,10 +37,11 @@
    * @param fileName the name of the file to be created
    * @param len file size in bytes
    * @param options options to create the file with
-   * @throws IOException if {@code path} is invalid (e.g., illegal URI)
+   * @throws IOException if an IO exception occurs
+   * @throws AlluxioException if an Alluxio exception occurs
    */
   public static void createByteFile(FileSystem fs, String fileName, int len,
-      CreateFileOptions options) throws IOException {
+      CreateFileOptions options) throws IOException, AlluxioException {
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/FileSystemTestUtils.java]: Throw exceptions in alphabetical order?

File: core/client/src/main/java/alluxio/client/FileSystemTestUtils.java [hunk 3]
```diff
@@ -37,10 +37,11 @@
    * @param fileName the name of the file to be created
    * @param len file size in bytes
    * @param options options to create the file with
-   * @throws IOException if {@code path} is invalid (e.g., illegal URI)
+   * @throws IOException if an IO exception occurs
+   * @throws AlluxioException if an Alluxio exception occurs
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/FileSystemTestUtils.java]: This javadoc comment seems redundant as well

File: core/common/src/main/java/alluxio/QuietlyCloseable.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ */
+
+package alluxio;
+
+import java.io.Closeable;
+
+/**
+ * A Closeable which doesn't throw any checked exceptions from its {@link #close()} method.
+ */
+public interface QuietlyCloseable extends Closeable {
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/QuietlyCloseable.java]: Should we provide a utility method called `closeQuietly`?

File: core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java [hunk 5]
```diff
@@ -117,42 +112,38 @@ public long remaining() {
...
-    if (throwable == null) {
+    if (exception == null) {
       mClosed = true;
       return;
     }
 
     try {
       mCloser.close();
-    } catch (Throwable e) {
+    } catch (Exception e) {
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java]: `CommonUtils.closeQuietly`?

File: core/client/src/main/java/alluxio/client/AbstractOutStream.java (hunk 1 above)
Comment (aaudiber) [inline core/client/src/main/java/alluxio/client/AbstractOutStream.java]: I'd prefer just `@throws IOException` without anything extra (this is what HDFS does), but I couldn't find a way to tell checkstyle that `IOException` and `AlluxioException` don't require javadoc descriptions. The alternative is to relax the check for all `@throws` clauses. I'd be OK with that, what do you think?

File: core/client/src/main/java/alluxio/client/FileSystemTestUtils.java (hunk 3 above)
Comment (aaudiber) [inline core/client/src/main/java/alluxio/client/FileSystemTestUtils.java]: same comment - I'd be fine with removing descriptions for `AlluxioException`

File: core/common/src/main/java/alluxio/QuietlyCloseable.java (hunk 4 above)
Comment (aaudiber) [inline core/common/src/main/java/alluxio/QuietlyCloseable.java]: It doesn't feel great to have only one of a class's methods declare that it throws `IOException` when it really doesn't. Can you explain more about your concerns towards adding a new interface?

File: core/client/src/main/java/alluxio/client/FileSystemTestUtils.java (hunk 2 above)
Comment (aaudiber) [inline core/client/src/main/java/alluxio/client/FileSystemTestUtils.java]: We use `IOException, AlluxioException` in 204 places and `AlluxioException, IOException` in 98 places >.< I'd prefer to punt alphabetization to a different PR

File: core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java (hunk 5 above)
Comment (aaudiber) [inline core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java]: `closeQuietly` only swallows `IOException`

**Comment by aaudiber** [comment]
@calvinjia Thanks for the quick high-level review

File: core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java (hunk 5 above)
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java]: Because we expect `close` to throw unchecked exceptions now? It might be worth updating our utility methods to swallow those as well, since that should be in line with the original purpose of the method.

File: core/client/src/main/java/alluxio/client/FileSystemTestUtils.java (hunk 2 above)
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/FileSystemTestUtils.java]: Yeah we have other places which need to go through this treatment, and we should enforce it in our checkstyle before attempting anything.

File: core/client/src/main/java/alluxio/client/AbstractOutStream.java (hunk 1 above)
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/AbstractOutStream.java]: I think we can relax the check for `@throws`, since a majority of these javadocs do not provide additional information. I would suggest to do this in another PR though.

File: core/common/src/main/java/alluxio/QuietlyCloseable.java (hunk 4 above)
Comment (calvinjia) [inline core/common/src/main/java/alluxio/QuietlyCloseable.java]: Is the motivation for this interface so that you can override `close` without throwing `IOException`? I think you can do that regardless of the interface throwing `IOException`. If we need the interface because at some point we need to distinguish between something implementing `Closeable` and `QuietlyCloseable`, I think it is cleaner to just have a utility method which does makes the `close` call equivalent to a `closequietly`.

File: core/client/src/main/java/alluxio/client/AbstractOutStream.java (hunk 1 above)
Comment (aaudiber) [inline core/client/src/main/java/alluxio/client/AbstractOutStream.java]: sgtm

File: core/common/src/main/java/alluxio/QuietlyCloseable.java (hunk 4 above)
Comment (aaudiber) [inline core/common/src/main/java/alluxio/QuietlyCloseable.java]: We discussed offline and decided that it makes sense to user `QuietlyCloseable` to create more precise interfaces which specify that they do not throw `IOExceptions`

File: core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java (hunk 5 above)
Comment (aaudiber) [inline core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java]: good point, I've updated `closeQuietly` and used it here

File: core/client/src/main/java/alluxio/client/FileSystemTestUtils.java (hunk 2 above)
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/FileSystemTestUtils.java]: should this class be in `test` instead of `main` if it is test utils? the reason I ask is because I would prefer to simplify the signature to `throws Exception` which is common-place in our test code base

File: core/client/src/main/java/alluxio/client/FileSystemTestUtils.java [hunk 6]
```diff
@@ -131,8 +133,8 @@ public static void createByteFile(FileSystem fs, String fileName,
         }
       }
       return res;
-    } catch (AlluxioException e) {
-      throw new IOException(e.getMessage());
+    } catch (IOException | AlluxioException e) {
+      throw new RuntimeException(e);
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/FileSystemTestUtils.java]: could we just do this for all of the utility methods?

File: core/client/src/main/java/alluxio/client/QuietlyCancelable.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+package alluxio.client;
+
+import java.io.IOException;
+
+/**
+ * Similar to {@link Cancelable}, but the cancel method doesn't throw {@link IOException}.
+ */
+public interface QuietlyCancelable {
+  /**
+   * Cancels the write to Alluxio storage. This will delete all the temporary data and metadata
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/QuietlyCancelable.java]: I think the method-level comment seems to be talking about implementation, which does not seem right

File: core/client/src/main/java/alluxio/client/block/stream/BlockInStream.java [hunk 8]
```diff
@@ -191,18 +178,15 @@ public static BlockInStream createUfsBlockInStream(FileSystemContext context, St
...
       CommonUtils.closeQuietly(closer);
       throw e;
     }
   }
 
   @Override
-  public void close() throws IOException {
-    mCloser.close();
+  public void close() {
+    CommonUtils.close(mCloser);
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/block/stream/BlockInStream.java]: why not `closeQuietly`?

File: core/client/src/main/java/alluxio/client/file/FileOutStream.java [hunk 9]
```diff
@@ -260,12 +283,12 @@ private long getNextBlockId() throws IOException {
     try (CloseableResource<FileSystemMasterClient> masterClient = mContext
         .acquireMasterClientResource()) {
       return masterClient.get().getNewBlockIdForFile(mUri);
-    } catch (AlluxioException e) {
+    } catch (AlluxioStatusException e) {
       throw new IOException(e);
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/FileOutStream.java]: this should either always be `e.toIOException()`, or we should not have that method

File: core/client/src/main/java/alluxio/client/file/FileOutStream.java [hunk 10]
```diff
@@ -280,13 +303,13 @@ private void handleCacheWriteException(IOException e) throws IOException {
...
-   * @throws IOException an I/O error occurs
+   * @throws IOException if an I/O error occurs
    */
   protected void scheduleAsyncPersist() throws IOException {
     try (CloseableResource<FileSystemMasterClient> masterClient = mContext
         .acquireMasterClientResource()) {
       masterClient.get().scheduleAsyncPersist(mUri);
-    } catch (AlluxioException e) {
+    } catch (AlluxioStatusException e) {
       throw new IOException(e);
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/FileOutStream.java]: ditto

File: core/client/src/main/java/alluxio/client/file/UnknownLengthFileInStream.java [hunk 11]
```diff
@@ -66,7 +66,7 @@ public void close() throws IOException {
       try {
         masterClient.completeFile(new AlluxioURI(mStatus.getPath()),
             CompleteFileOptions.defaults().setUfsLength(mPos));
-      } catch (AlluxioException e) {
+      } catch (AlluxioStatusException e) {
         throw new IOException(e);
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/UnknownLengthFileInStream.java]: ditto

File: core/client/src/main/java/alluxio/client/lineage/AbstractLineageClient.java [hunk 12]
```diff
@@ -63,12 +65,19 @@ public long createLineage(List<AlluxioURI> inputFiles, List<AlluxioURI> outputFi
...
     try {
+      masterClient = mContext.acquireMasterClient();
       long lineageId = masterClient.createLineage(stripURIList(inputFiles),
           stripURIList(outputFiles), (CommandLineJob) job);
       LOG.info("Created lineage {}", lineageId);
       return lineageId;
+    } catch (NotFoundException e) {
+      throw new FileDoesNotExistException(e.getMessage());
+    } catch (UnavailableException e) {
+      throw new IOException(e.getMessage());
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/lineage/AbstractLineageClient.java]: ditto

File: core/client/src/main/java/alluxio/client/lineage/AbstractLineageClient.java [hunk 13]
```diff
@@ -63,12 +65,19 @@ public long createLineage(List<AlluxioURI> inputFiles, List<AlluxioURI> outputFi
...
+    } catch (AlluxioStatusException e) {
+      throw e.toAlluxioException();
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/lineage/AbstractLineageClient.java]: `e.toIOException`?

File: core/client/src/main/java/alluxio/client/lineage/AbstractLineageClient.java [hunk 14]
```diff
@@ -77,11 +86,16 @@ public long createLineage(List<AlluxioURI> inputFiles, List<AlluxioURI> outputFi
...
+    LineageMasterClient masterClient = null;
     try {
+      masterClient = mContext.acquireMasterClient();
       boolean result = masterClient.deleteLineage(lineageId, options.isCascade());
       LOG.info("{} delete lineage {}", result ? "Succeeded to " : "Failed to ", lineageId);
       return result;
+    } catch (NotFoundException e) {
+      throw new LineageDoesNotExistException(e.getMessage());
+    } catch (AlluxioStatusException e) {
+      throw new LineageDeletionException(e.getMessage());
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/lineage/AbstractLineageClient.java]: `e.toIOException`?

File: core/client/src/main/java/alluxio/client/lineage/LineageFileSystem.java [hunk 15]
```diff
@@ -67,6 +69,12 @@ private long reinitializeFile(AlluxioURI path, CreateFileOptions options)
     try {
       return masterClient.reinitializeFile(path.getPath(), options.getBlockSizeBytes(),
           options.getTtl(), options.getTtlAction());
+    } catch (NotFoundException e) {
+      throw new LineageDoesNotExistException(e.getMessage());
+    } catch (UnavailableException e) {
+      throw new IOException(e.getMessage());
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/lineage/LineageFileSystem.java]: `e.toIOException`?

File: core/client/src/main/java/alluxio/client/lineage/LineageFileSystem.java [hunk 16]
```diff
@@ -111,6 +119,12 @@ public void reportLostFile(AlluxioURI path)
     LineageMasterClient masterClient = mLineageContext.acquireMasterClient();
     try {
       masterClient.reportLostFile(path.getPath());
+    } catch (NotFoundException e) {
+      throw new FileDoesNotExistException(e.getMessage());
+    } catch (UnavailableException e) {
+      throw new IOException(e.getMessage());
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/lineage/LineageFileSystem.java]: ditto

File: core/client/src/test/java/alluxio/client/file/FileInStreamTest.java [hunk 17]
```diff
@@ -433,7 +434,7 @@ public void testPromote() throws IOException {
   @Test
   public void failGetInStream() throws IOException {
     Mockito.when(mBlockStore.getInStream(Mockito.eq(1L), Mockito.any(InStreamOptions.class)))
-        .thenThrow(new IOException("test IOException"));
+        .thenThrow(new UnavailableException("test IOException"));
```
Comment (jsimsa) [inline core/client/src/test/java/alluxio/client/file/FileInStreamTest.java]: change `IOException` to `exception`

File: core/client/src/test/java/alluxio/client/file/FileInStreamTest.java [hunk 18]
```diff
@@ -455,7 +456,7 @@ public void failToUnderFs() throws AlluxioException, IOException {
             mContext);
 
     Mockito.when(mBlockStore.getInStream(Mockito.eq(1L), Mockito.any(InStreamOptions.class)))
-        .thenThrow(new IOException("test IOException"));
+        .thenThrow(new UnavailableException("test IOException"));
```
Comment (jsimsa) [inline core/client/src/test/java/alluxio/client/file/FileInStreamTest.java]: ditto

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 19]
```diff
@@ -283,95 +281,40 @@ public synchronized InetSocketAddress getAddress() {
   }
 
   /**
-   * Same with {@link RpcCallable} except that this RPC call throws {@link AlluxioTException} and
-   * is to be executed in {@link #retryRPC(RpcCallableThrowsAlluxioTException)}.
-   *
-   * @param <V> the return value of {@link #call()}
-   */
-  protected interface RpcCallableThrowsAlluxioTException<V> {
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/AbstractClient.java]: \o/

File: core/common/src/main/java/alluxio/QuietlyCloseable.java (hunk 4 above)
Comment (jsimsa) [inline core/common/src/main/java/alluxio/QuietlyCloseable.java]: FYI, instead of introducing `QuietlyClosable` you can change redeclared the method again in an interface that extends `Closable` with a more strict signature

Comment (jsimsa) [inline core/common/src/main/java/alluxio/QuietlyCloseable.java]: ditto for `QuietlyCancelable`

File: core/common/src/main/java/alluxio/exception/status/AlluxioStatusException.java [hunk 20]
```diff
@@ -116,8 +126,19 @@ public AlluxioException toAlluxioException() {
...
-        return new AlluxioException(getMessage());
+        return new AlluxioException(getMessage(), this);
+    }
+  }
+
+  /**
+   * @return this exception converted to an IOException
+   */
+  public IOException toIOException() {
+    // If this exception was
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/exception/status/AlluxioStatusException.java]: runaway comment?

File: keyvalue/client/src/main/java/alluxio/client/keyvalue/BaseKeyValueSystem.java [hunk 21]
```diff
@@ -44,30 +47,62 @@ public BaseKeyValueSystem() {}
   @Override
   public KeyValueStoreReader openStore(AlluxioURI uri) throws IOException, AlluxioException {
     Preconditions.checkNotNull(uri, PreconditionMessage.URI_KEY_VALUE_STORE_NULL);
-    return new BaseKeyValueStoreReader(uri);
+    try {
+      return new BaseKeyValueStoreReader(uri);
+    } catch (UnavailableException e) {
+      throw new IOException(e);
```
Comment (jsimsa) [inline keyvalue/client/src/main/java/alluxio/client/keyvalue/BaseKeyValueSystem.java]: `e.toAlluxioException()`?

File: keyvalue/client/src/main/java/alluxio/hadoop/mapreduce/KeyValueInputSplit.java [hunk 22]
```diff
@@ -74,7 +74,7 @@ public long getLength() throws IOException {
         locations[i] = workersInfo.get(i).getNetAddress().getHost();
       }
       return locations;
-    } catch (AlluxioException e) {
+    } catch (AlluxioStatusException e) {
       throw new IOException(e);
```
Comment (jsimsa) [inline keyvalue/client/src/main/java/alluxio/hadoop/mapreduce/KeyValueInputSplit.java]: `e.toIOException()`

File: core/client/src/main/java/alluxio/client/lineage/AbstractLineageClient.java (hunk 14 above)
Comment (aaudiber) [inline core/client/src/main/java/alluxio/client/lineage/AbstractLineageClient.java]: changed to convert these a little better, probably not very important since I doubt anyone in the world is handling these exceptions differently.

File: core/client/src/main/java/alluxio/client/lineage/LineageFileSystem.java (hunk 15 above)
Comment (aaudiber) [inline core/client/src/main/java/alluxio/client/lineage/LineageFileSystem.java]: done

File: core/client/src/main/java/alluxio/client/lineage/LineageFileSystem.java (hunk 16 above)
Comment (aaudiber) [inline core/client/src/main/java/alluxio/client/lineage/LineageFileSystem.java]: done

File: core/client/src/test/java/alluxio/client/file/FileInStreamTest.java (hunk 17 above)
Comment (aaudiber) [inline core/client/src/test/java/alluxio/client/file/FileInStreamTest.java]: done

File: core/client/src/test/java/alluxio/client/file/FileInStreamTest.java (hunk 18 above)
Comment (aaudiber) [inline core/client/src/test/java/alluxio/client/file/FileInStreamTest.java]: done

File: core/common/src/main/java/alluxio/QuietlyCloseable.java (hunk 4 above)
Comment (aaudiber) [inline core/common/src/main/java/alluxio/QuietlyCloseable.java]: I'm currently having 8 interfaces and one class extend/implement `QuietlyCloseable` so that they don't need to repeat its content. I don't see much advantage in repeating in all of these places. `QuietlyCancelable` is mostly for consistency with `QuietlyCloseable`, it's used for just 1 interface and 2 classes.

File: core/common/src/main/java/alluxio/exception/status/AlluxioStatusException.java (hunk 20 above)
Comment (aaudiber) [inline core/common/src/main/java/alluxio/exception/status/AlluxioStatusException.java]: oops, fixed

File: core/client/src/test/java/alluxio/client/file/FileSystemTestUtils.java [hunk 23]
```diff
@@ -9,9 +9,11 @@
  * See the NOTICE file distributed with this work for information regarding copyright ownership.
  */
 
-package alluxio.client;
+package alluxio.client.file;
 
 import alluxio.AlluxioURI;
+import alluxio.client.ReadType;
+import alluxio.client.WriteType;
 import alluxio.client.file.FileOutStream;
```
Comment (jsimsa) [inline core/client/src/test/java/alluxio/client/file/FileSystemTestUtils.java]: do you still need to import `alluxio.client.file.*` now that it has been moved to the package?

Comment (jsimsa) [inline core/client/src/test/java/alluxio/client/file/FileSystemTestUtils.java]: nevermind, I noticed they are removed in another commit

File: core/client/src/main/java/alluxio/client/lineage/AbstractLineageClient.java (hunk 13 above)
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/lineage/AbstractLineageClient.java]: why do we sometimes convert `AlluxioStatusException` to `IOException` (e.g. in `FileOutStream.java`) and other times to `AlluxioException` (e.g. here)?

**Comment by jsimsa** [comment]
@aaudiber thank you for the interation, did another pass and left one more question ... otherwise LGTM

File: keyvalue/client/src/main/java/alluxio/client/keyvalue/BaseKeyValueSystem.java (hunk 21 above)
Comment (aaudiber) [inline keyvalue/client/src/main/java/alluxio/client/keyvalue/BaseKeyValueSystem.java]: done

File: keyvalue/client/src/main/java/alluxio/hadoop/mapreduce/KeyValueInputSplit.java (hunk 22 above)
Comment (aaudiber) [inline keyvalue/client/src/main/java/alluxio/hadoop/mapreduce/KeyValueInputSplit.java]: done

**Comment by aaudiber** [comment]
@calvinjia @jsimsa Comments addressed, PTAL

File: core/client/src/main/java/alluxio/client/lineage/AbstractLineageClient.java (hunk 13 above)
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/lineage/AbstractLineageClient.java]: ping

Comment (aaudiber) [inline core/client/src/main/java/alluxio/client/lineage/AbstractLineageClient.java]: sorry, missed this comment

For users `IOException` usually means "Exception that might be worth retrying", which is also the meaning of `UnavailableException`. Therefore, we convert `UnavailableException` to `IOException` to preserve this. Other types of status exception are generally not worth retrying, so we convert them to `AlluxioException`, which previously meant "something that went wrong in Alluxio that isn't IOException"

Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/lineage/AbstractLineageClient.java]: what I am pointing out is that there does not seem to be a consistent pattern for what to convert `AlluxioStatusException` to ... which means that it will be inconsistent

I assume this inconsistency will go away in 2.0 when we will be able to change the user API, correct?

Comment (aaudiber) [inline core/client/src/main/java/alluxio/client/lineage/AbstractLineageClient.java]: There should be a consistent pattern:
```java
try {
  operation()
} catch (SpecificStatusExceptionCorrespondingToThrownSubclassOfAlluxioExceptionForThisSpecificMethod e) {
  throw new ThrownSubclassOfAlluxioExceptionForThisSpecificMethod1(e.getMessage());
...
} catch (UnavailableException e) {
  throw e.toIOException();
} catch (AlluxioStatusException e) {
  throw e.toAlluxioException();
}
```

For example, 

```java
  @Override
  public boolean deleteLineage(long lineageId, DeleteLineageOptions options)
      throws IOException, LineageDoesNotExistException, LineageDeletionException, AlluxioException {
    LineageMasterClient masterClient = null;
    try {
      masterClient = mContext.acquireMasterClient();
      boolean result = masterClient.deleteLineage(lineageId, options.isCascade());
      LOG.info("{} delete lineage {}", result ? "Succeeded to " : "Failed to ", lineageId);
      return result;
    } catch (NotFoundException e) {
      throw new LineageDoesNotExistException(e.getMessage());
    } catch (FailedPreconditionException e) {
      throw new LineageDeletionException(e.getMessage());
    } catch (UnavailableException e) {
      throw e.toIOException();
    } catch (AlluxioStatusException e) {
      throw e.toAlluxioException();
    } finally {
      mContext.releaseMasterClient(masterClient);
    }
  }
```

I forgot to handle `UnavailableException` in `BaseFileSystem`, it's updated now.

After 2.0 all of this will go away and we won't catch anything  at the public API boundary

**Comment by aaudiber** [comment]
Jenkins, test this please
